homeassistant/components/metoffice/sensor.py

```diff
@@ -30,175 +32,163 @@
+                if self.metoffice_now.weather.value in v
+            ][0]
+
+        elif hasattr(self.metoffice_now, self._type):
+            value = getattr(self.metoffice_now, self._type)
+
+            if not isinstance(value, int):
+                value = value.value
+
+        return value
 
     @property
     def unit_of_measurement(self):
         """Return the unit of measurement."""
-        return SENSOR_TYPES[self._condition][1]
+        return SENSOR_TYPES[self._type][1]
+
+    @property
+    def icon(self):
+        """Return the icon for the entity card."""
+        value = SENSOR_TYPES[self._type][2]
+        if self._type == "weather":
+            value = self.state
+            if value == "partlycloudy":
+                value = "partly-cloudy"
+            value = f"mdi:weather-{value}"
+
+        return value
 
     @property
     def device_state_attributes(self):
         """Return the state attributes of the device."""
-        attr = {}
-        attr[ATTR_ATTRIBUTION] = ATTRIBUTION
-        attr[ATTR_LAST_UPDATE] = self.data.data.date
-        attr[ATTR_SENSOR_ID] = self._condition
-        attr[ATTR_SITE_ID] = self.site.id
-        attr[ATTR_SITE_NAME] = self.site.name
-        return attr
-
-    def update(self):
-        """Update current conditions."""
-        self.data.update()
-
-
-class MetOfficeCurrentData:
-    """Get data from Datapoint."""
-
-    def __init__(self, hass, datapoint, site):
-        """Initialize the data object."""
-        self._datapoint = datapoint
-        self._site = site
-        self.data = None
-
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Get the latest data from Datapoint."""
-        try:
-            forecast = self._datapoint.get_forecast_for_site(self._site.id, "3hourly")
-            self.data = forecast.now()
-        except (ValueError, dp.exceptions.APIException) as err:
-            _LOGGER.error("Check Met Office %s", err.args)
-            self.data = None
+        return {
+            ATTR_ATTRIBUTION: ATTRIBUTION,
+            ATTR_LAST_UPDATE: self.metoffice_now.date if self.metoffice_now else None,
+            ATTR_SENSOR_ID: self._type,
+            ATTR_SITE_ID: self.metoffice_site_id if self.metoffice_site_id else None,
+            ATTR_SITE_NAME: self.metoffice_site_name
+            if self.metoffice_site_name
+            else None,
+        }
+
+    async def async_added_to_hass(self) -> None:
+        """Set up a listener and load data."""
+        self.async_on_remove(
+            self._coordinator.async_add_listener(self._update_callback)
+        )
+        self._update_callback()
+
+    @callback
+    def _update_callback(self) -> None:
+        """Load data from integration."""
+        self.metoffice_site_id = self._data.site_id
+        self.metoffice_site_name = self._data.site_name
+        self.metoffice_now = self._data.now
+        self.async_write_ha_state()
+
```

Comment:
Please also implement `async_update` method and await `coordinator.async_request_refresh`. That will allow users to schedule custom updates via the common entity update service.

Reply:
Done.